zebra-chain/src/sapling/shielded_data.rs

```diff
@@ -216,39 +273,60 @@ where
     }
 }
 
-// Technically, it's possible to construct two equivalent representations
-// of a ShieldedData with at least one spend and at least one output, depending
-// on which goes in the `first` slot.  This is annoying but a smallish price to
-// pay for structural validity.
-//
-// A `ShieldedData<PerSpendAnchor>` can never be equal to a
-// `ShieldedData<SharedAnchor>`, even if they have the same effects.
-
-impl<AnchorV> std::cmp::PartialEq for ShieldedData<AnchorV>
+impl<AnchorV> TransferData<AnchorV>
 where
-    AnchorV: AnchorVariant + Clone + PartialEq,
+    AnchorV: AnchorVariant + Clone,
 {
-    fn eq(&self, other: &Self) -> bool {
-        // First check that the lengths match, so we know it is safe to use zip,
-        // which truncates to the shorter of the two iterators.
-        if self.spends().count() != other.spends().count() {
-            return false;
-        }
-        if self.outputs().count() != other.outputs().count() {
-            return false;
+    /// Iterate over the [`Spend`]s for this transaction, returning them as
+    /// their generic type.
+    pub fn spends(&self) -> impl Iterator<Item = &Spend<AnchorV>> {
+        use TransferData::*;
+
+        // TODO: replace with Box<dyn IntoIterator ...> for efficiency?
+        match self {
+            Spends {
+                first_spend,
+                rest_spends,
+                ..
+            } => iter::once(first_spend).chain(rest_spends.iter()).collect(),
+            NoSpends { .. } => Vec::new(),
         }
+        .into_iter()
+    }
+
+    /// Iterate over the [`Output`]s for this transaction.
+    pub fn outputs(&self) -> impl Iterator<Item = &Output> {
+        use TransferData::*;
+
+        // TODO: replace with Box<dyn IntoIterator ...> for efficiency?
+        let outputs: Vec<_> = match self {
+            Spends { maybe_outputs, .. } => maybe_outputs.iter().collect(),
+            NoSpends {
+                first_output,
+                rest_outputs,
+                ..
+            } => iter::once(first_output)
+                .chain(rest_outputs.iter())
+                .collect(),
```

Comment:
Actually we should just make one iterator using `<Option as IntoIterator<...>>`:
```rust
        match self.first {
            Either::Left(_) => None,
            Either::Right(ref output) => Some(output),
        }
        .into_iter()
        .chain(self.rest_outputs.iter())